{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import itertools\n",
    "import os\n",
    "\n",
    "import similarityregression.PairwiseAlignment as pwsaln\n",
    "import similarityregression.AlignmentTools as alntools\n",
    "import similarityregression.PredictSimilarity as srpred"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Introduction\n",
    "\n",
    "This notebook describes an example of how to parse protein sequences from CisBP into a form that can be used for Similarity Regression (training and scoring). Source data: 5 Homeodomain PBM experiments from CisBP (http://cisbp.ccbr.utoronto.ca/). Motif Similarity was calculated with MoSBAT (http://mosbat.ccbr.utoronto.ca/) as an example (E-score overlaps are used in the paper)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Parse CisBP Data\n",
    "## Read E-scores and calculate motif similarity (E-Score Overlaps)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "escores = pd.read_csv('HomeodomainData/EScore.txt', index_col=0, delimiter='\\t')\n",
    "escores.columns = [x.split(':')[0] for x in escores.columns] #Relabel by moits\n",
    "escores = escores.groupby(by=escores.columns, axis=1).mean() #Average E-scores for replicates\n",
    "\n",
    "#Calculate E-score Overlaps\n",
    "scorethresh = 0.45\n",
    "EScoreOverlaps = {}\n",
    "\n",
    "MIDs = list(escores.columns)\n",
    "MIDs.sort()\n",
    "for x,y in itertools.combinations(MIDs, 2):\n",
    "    #print x, y\n",
    "    maxN = max(len(set(escores[escores[x] >= scorethresh][x].index)), len(set(escores[escores[y] >= scorethresh][y].index)))\n",
    "    x_maxN = set(escores[x].sort_values(ascending = False).iloc[:maxN].index)\n",
    "    y_maxN = set(escores[y].sort_values(ascending = False).iloc[:maxN].index)\n",
    "    escoreoverlap = 1.0*len(x_maxN.intersection(y_maxN))/maxN\n",
    "    #print escoreoverlap, len(x_maxN.intersection(y_maxN)), maxN\n",
    "    EScoreOverlaps[(x,y)] = escoreoverlap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# TF / Protein Information \n",
    "## Parse Protein Information to reference DBD alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "proteins = pd.read_csv('HomeodomainData/prot_seq.txt', index_col=0, delimiter='\\t')\n",
    "\n",
    "motifs = []\n",
    "with open('HomeodomainData/TF_Information.txt', 'r') as infile:\n",
    "    h = infile.readline().strip().split('\\t')\n",
    "    for line in infile:\n",
    "        line = line.strip().split('\\t')\n",
    "        data = line[:6]\n",
    "        mids = line[6:]\n",
    "        mids = [x[:-1] for x in mids]\n",
    "        for mid in mids:\n",
    "            motifs.append([mid] + data)\n",
    "motifs = pd.DataFrame(motifs, columns=[h[-1]] + h[:-1])\n",
    "motifs.set_index('Motif_ID', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "DBDseqs = set()\n",
    "for x in proteins['DBD_seqs']:\n",
    "    x = x.split(',') #in cases of multiple DBDs\n",
    "    for seq in x:\n",
    "        DBDseqs.add(seq)\n",
    "        \n",
    "#Write DBDseqs to fasta\n",
    "with open('HDSeqs.fa', 'w') as outfile:\n",
    "    for DBDseq in DBDseqs:\n",
    "        outfile.write('>' + DBDseq + '\\n' + DBDseq + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Align to Pfam HMM (source: https://pfam.xfam.org/family/PF00046/hmm)\n",
    "\n",
    "`python RunAPHID.py ../Example/HomeodomainData/Homeodomain.hmm ./Example/HDSeqs.fa semiglobal`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#Read results into dictionary that maps the DBD sequence to its reference alignment\n",
    "DBDseqs = {} #Actual DBDseq : Reference Alignment \n",
    "for seq, aln in alntools.FastaIter(fileloc='DBDMatchPos_aphid/HDSeqs.matchpos_semiglobal.fa'):\n",
    "    DBDseqs[seq] = aln"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Get Motif Sequences and Alignments to Reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "MotifSequences = {} # MID: [DBD Sequences]\n",
    "for mid in MIDs:\n",
    "    #print mid\n",
    "    minfo = motifs.loc[mid]\n",
    "    protinfo = proteins.loc[minfo['TF_ID']]\n",
    "    if type(protinfo['DBD_seqs']) == str:\n",
    "        MotifSequences[mid] = protinfo['DBD_seqs'].split(',')\n",
    "    else:\n",
    "        MotifSequences[mid] = protinfo['DBD_seqs'][0].split(',')\n",
    "    \n",
    "#Get Motifs Sequences that are aligned to common Pfam reference\n",
    "MotifSequences_aligned = {}\n",
    "for mid, unaligned in MotifSequences.items():\n",
    "    MotifSequences_aligned[mid] = [DBDseqs.get(x) for x in unaligned]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Make DBD alignments and positional features for motif pairs with E-score overlaps\n",
    "\n",
    "- The MotifAlignments contain the features, and outputs necessary to output into dataframes for training in R. \n",
    "- They can also be scored using existing SR models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "Homeodomain_ReplicateThreshold = 0.5555\n",
    "\n",
    "MotifAlignments = {} # MID Pair : {Alignment Information}\n",
    "\n",
    "for pair, overlap in EScoreOverlaps.items():\n",
    "    #Get Reference alignments for each motif\n",
    "    proteinseqs_x = (pair[0], MotifSequences_aligned[pair[0]])\n",
    "    proteinseqs_y = (pair[1], MotifSequences_aligned[pair[1]])\n",
    "    #Align each motif to eachother\n",
    "    sr_alignment = pwsaln.AlignDBDArrays(proteinseqs_x, proteinseqs_y)\n",
    "    #Add E-score overlap to the alignment information\n",
    "    sr_alignment['EScoreOverlap'] = overlap\n",
    "    sr_alignment['EClass'] = int(overlap >= Homeodomain_ReplicateThreshold)\n",
    "    MotifAlignments[pair] = sr_alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Examples of information in an SR alignment \n",
    "Example: \n",
    "\n",
    "M1009_1.02:Berger08:Hoxa7_2668 \n",
    "\n",
    "    vs.\t\n",
    "\n",
    "M1072_1.02:Badis09:Hoxa3_2783"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Motif Similarity (E-score Overlap): 0.637931034483\n",
      "Alignment % Amino Acid Identity: 73.684%\n",
      "Positional Amino Acid Identity: [1.0, 1.0, 0.0, 1.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 1.0, 0.0, 1.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 1.0, 1.0]\n",
      "Positional Amino Acid Similarity: [5.0, 5.0, 0.0, 5.0, -1.0, 0.0, 7.0, 5.0, -1.0, -2.0, 5.0, -1.0, 1.0, 5.0, 4.0, 5.0, 5.0, 5.0, 6.0, 8.0, 6.0, 6.0, 5.0, 7.0, 4.0, -1.0, 5.0, -2.0, 5.0, 5.0, 3.0, 5.0, 1.0, 4.0, 1.0, -1.0, 4.0, -3.0, 4.0, 5.0, 5.0, 5.0, 5.0, 4.0, 5.0, 4.0, 11.0, 6.0, 5.0, 6.0, 5.0, 5.0, 5.0, 5.0, 2.0, 5.0, 5.0]\n"
     ]
    }
   ],
   "source": [
    "sr_alignment = MotifAlignments[('M1009_1.02', 'M1072_1.02')]\n",
    "\n",
    "print 'Motif Similarity (E-score Overlap):', sr_alignment['EScoreOverlap']\n",
    "print 'Alignment % Amino Acid Identity:', '{:.3f}'.format(100*sr_alignment['PctID_L']) + '%'\n",
    "print 'Positional Amino Acid Identity:', sr_alignment['ByPos.PctID']\n",
    "print 'Positional Amino Acid Similarity:', sr_alignment['ByPos.AvgB62']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Example of Scoring TF pairs using an existing model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>AA %ID</th>\n",
       "      <th>EScoreOverlap</th>\n",
       "      <th>SR_Score</th>\n",
       "      <th>SR_Class</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M1072_1.02</th>\n",
       "      <th>M1073_1.02</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.887640</td>\n",
       "      <td>0.866518</td>\n",
       "      <td>HSim</td>\n",
       "      <td>HSim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1009_1.02</th>\n",
       "      <th>M1010_1.02</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.775000</td>\n",
       "      <td>0.866518</td>\n",
       "      <td>HSim</td>\n",
       "      <td>HSim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">M1010_1.02</th>\n",
       "      <th>M1073_1.02</th>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.623596</td>\n",
       "      <td>0.622210</td>\n",
       "      <td>HSim</td>\n",
       "      <td>HSim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1072_1.02</th>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.597701</td>\n",
       "      <td>0.622210</td>\n",
       "      <td>HSim</td>\n",
       "      <td>HSim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">M1009_1.02</th>\n",
       "      <th>M1073_1.02</th>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.651685</td>\n",
       "      <td>0.622210</td>\n",
       "      <td>HSim</td>\n",
       "      <td>HSim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1072_1.02</th>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.637931</td>\n",
       "      <td>0.622210</td>\n",
       "      <td>HSim</td>\n",
       "      <td>HSim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">M0931_1.02</th>\n",
       "      <th>M1073_1.02</th>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.643192</td>\n",
       "      <td>0.549141</td>\n",
       "      <td>Amb</td>\n",
       "      <td>HSim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1072_1.02</th>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.671362</td>\n",
       "      <td>0.549141</td>\n",
       "      <td>Amb</td>\n",
       "      <td>HSim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">M1007_1.02</th>\n",
       "      <th>M1073_1.02</th>\n",
       "      <td>0.526316</td>\n",
       "      <td>0.320225</td>\n",
       "      <td>0.522381</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Amb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1072_1.02</th>\n",
       "      <td>0.526316</td>\n",
       "      <td>0.293103</td>\n",
       "      <td>0.522381</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Amb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1010_1.02</th>\n",
       "      <td>0.526316</td>\n",
       "      <td>0.410526</td>\n",
       "      <td>0.520482</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Amb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1009_1.02</th>\n",
       "      <td>0.526316</td>\n",
       "      <td>0.383333</td>\n",
       "      <td>0.520482</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Amb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">M0966_1.02</th>\n",
       "      <th>M1007_1.02</th>\n",
       "      <td>0.456140</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.474253</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1009_1.02</th>\n",
       "      <td>0.421053</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.453149</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1010_1.02</th>\n",
       "      <td>0.421053</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.453149</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1072_1.02</th>\n",
       "      <td>0.456140</td>\n",
       "      <td>0.005747</td>\n",
       "      <td>0.420110</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1073_1.02</th>\n",
       "      <td>0.456140</td>\n",
       "      <td>0.005618</td>\n",
       "      <td>0.420110</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">M0931_1.02</th>\n",
       "      <th>M1007_1.02</th>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.342723</td>\n",
       "      <td>0.406922</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Amb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0966_1.02</th>\n",
       "      <td>0.298246</td>\n",
       "      <td>0.004695</td>\n",
       "      <td>0.404151</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">M0941_1.02</th>\n",
       "      <th>M1073_1.02</th>\n",
       "      <td>0.491228</td>\n",
       "      <td>0.005618</td>\n",
       "      <td>0.398389</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1072_1.02</th>\n",
       "      <td>0.491228</td>\n",
       "      <td>0.005747</td>\n",
       "      <td>0.398389</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">M0931_1.02</th>\n",
       "      <th>M1010_1.02</th>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.624413</td>\n",
       "      <td>0.378838</td>\n",
       "      <td>Amb</td>\n",
       "      <td>HSim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1009_1.02</th>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.624413</td>\n",
       "      <td>0.378838</td>\n",
       "      <td>Amb</td>\n",
       "      <td>HSim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0941_1.02</th>\n",
       "      <td>0.385965</td>\n",
       "      <td>0.028169</td>\n",
       "      <td>0.362176</td>\n",
       "      <td>Amb</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0909_1.02</th>\n",
       "      <th>M0933_1.02</th>\n",
       "      <td>0.381818</td>\n",
       "      <td>0.042056</td>\n",
       "      <td>0.238042</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">M0941_1.02</th>\n",
       "      <th>M1009_1.02</th>\n",
       "      <td>0.456140</td>\n",
       "      <td>0.006098</td>\n",
       "      <td>0.230076</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1010_1.02</th>\n",
       "      <td>0.456140</td>\n",
       "      <td>0.006098</td>\n",
       "      <td>0.230076</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1049_1.02</th>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.199632</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1007_1.02</th>\n",
       "      <td>0.368421</td>\n",
       "      <td>0.006098</td>\n",
       "      <td>0.183438</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0966_1.02</th>\n",
       "      <td>0.368421</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.155471</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">M0933_1.02</th>\n",
       "      <th>M0966_1.02</th>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.004673</td>\n",
       "      <td>0.128590</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1007_1.02</th>\n",
       "      <td>0.298246</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.090370</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1010_1.02</th>\n",
       "      <td>0.210526</td>\n",
       "      <td>0.004673</td>\n",
       "      <td>0.087494</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1009_1.02</th>\n",
       "      <td>0.210526</td>\n",
       "      <td>0.004673</td>\n",
       "      <td>0.087494</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">M0909_1.02</th>\n",
       "      <th>M1009_1.02</th>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.085192</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1010_1.02</th>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.085192</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0966_1.02</th>\n",
       "      <td>0.280702</td>\n",
       "      <td>0.007576</td>\n",
       "      <td>0.063947</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0933_1.02</th>\n",
       "      <th>M1049_1.02</th>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.054308</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0931_1.02</th>\n",
       "      <th>M0933_1.02</th>\n",
       "      <td>0.290909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054079</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0933_1.02</th>\n",
       "      <th>M0941_1.02</th>\n",
       "      <td>0.280702</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.049819</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">M0909_1.02</th>\n",
       "      <th>M1007_1.02</th>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.049280</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1049_1.02</th>\n",
       "      <td>0.381818</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.043178</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0941_1.02</th>\n",
       "      <td>0.280702</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.042790</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">M0933_1.02</th>\n",
       "      <th>M1073_1.02</th>\n",
       "      <td>0.192982</td>\n",
       "      <td>0.004673</td>\n",
       "      <td>0.042592</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1072_1.02</th>\n",
       "      <td>0.192982</td>\n",
       "      <td>0.004673</td>\n",
       "      <td>0.042592</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">M0909_1.02</th>\n",
       "      <th>M1072_1.02</th>\n",
       "      <td>0.245614</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.036407</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1073_1.02</th>\n",
       "      <td>0.245614</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.036407</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0931_1.02</th>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.009390</td>\n",
       "      <td>0.023920</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0931_1.02</th>\n",
       "      <th>M1049_1.02</th>\n",
       "      <td>0.309091</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.027452</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1009_1.02</th>\n",
       "      <th>M1049_1.02</th>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.027919</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1010_1.02</th>\n",
       "      <th>M1049_1.02</th>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.027919</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1007_1.02</th>\n",
       "      <th>M1049_1.02</th>\n",
       "      <td>0.218182</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.036610</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">M1049_1.02</th>\n",
       "      <th>M1072_1.02</th>\n",
       "      <td>0.228070</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.039781</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1073_1.02</th>\n",
       "      <td>0.228070</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.039781</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0966_1.02</th>\n",
       "      <th>M1049_1.02</th>\n",
       "      <td>0.290909</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.045417</td>\n",
       "      <td>Dis</td>\n",
       "      <td>Dis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         AA %ID  EScoreOverlap  SR_Score SR_Class Class\n",
       "M1072_1.02 M1073_1.02  1.000000       0.887640  0.866518     HSim  HSim\n",
       "M1009_1.02 M1010_1.02  1.000000       0.775000  0.866518     HSim  HSim\n",
       "M1010_1.02 M1073_1.02  0.736842       0.623596  0.622210     HSim  HSim\n",
       "           M1072_1.02  0.736842       0.597701  0.622210     HSim  HSim\n",
       "M1009_1.02 M1073_1.02  0.736842       0.651685  0.622210     HSim  HSim\n",
       "           M1072_1.02  0.736842       0.637931  0.622210     HSim  HSim\n",
       "M0931_1.02 M1073_1.02  0.333333       0.643192  0.549141      Amb  HSim\n",
       "           M1072_1.02  0.333333       0.671362  0.549141      Amb  HSim\n",
       "M1007_1.02 M1073_1.02  0.526316       0.320225  0.522381      Amb   Amb\n",
       "           M1072_1.02  0.526316       0.293103  0.522381      Amb   Amb\n",
       "           M1010_1.02  0.526316       0.410526  0.520482      Amb   Amb\n",
       "           M1009_1.02  0.526316       0.383333  0.520482      Amb   Amb\n",
       "M0966_1.02 M1007_1.02  0.456140       0.000000  0.474253      Amb   Dis\n",
       "           M1009_1.02  0.421053       0.000000  0.453149      Amb   Dis\n",
       "           M1010_1.02  0.421053       0.000000  0.453149      Amb   Dis\n",
       "           M1072_1.02  0.456140       0.005747  0.420110      Amb   Dis\n",
       "           M1073_1.02  0.456140       0.005618  0.420110      Amb   Dis\n",
       "M0931_1.02 M1007_1.02  0.315789       0.342723  0.406922      Amb   Amb\n",
       "           M0966_1.02  0.298246       0.004695  0.404151      Amb   Dis\n",
       "M0941_1.02 M1073_1.02  0.491228       0.005618  0.398389      Amb   Dis\n",
       "           M1072_1.02  0.491228       0.005747  0.398389      Amb   Dis\n",
       "M0931_1.02 M1010_1.02  0.333333       0.624413  0.378838      Amb  HSim\n",
       "           M1009_1.02  0.333333       0.624413  0.378838      Amb  HSim\n",
       "           M0941_1.02  0.385965       0.028169  0.362176      Amb   Dis\n",
       "M0909_1.02 M0933_1.02  0.381818       0.042056  0.238042      Dis   Dis\n",
       "M0941_1.02 M1009_1.02  0.456140       0.006098  0.230076      Dis   Dis\n",
       "           M1010_1.02  0.456140       0.006098  0.230076      Dis   Dis\n",
       "           M1049_1.02  0.327273       0.000000  0.199632      Dis   Dis\n",
       "           M1007_1.02  0.368421       0.006098  0.183438      Dis   Dis\n",
       "           M0966_1.02  0.368421       0.000000  0.155471      Dis   Dis\n",
       "M0933_1.02 M0966_1.02  0.315789       0.004673  0.128590      Dis   Dis\n",
       "           M1007_1.02  0.298246       0.000000  0.090370      Dis   Dis\n",
       "           M1010_1.02  0.210526       0.004673  0.087494      Dis   Dis\n",
       "           M1009_1.02  0.210526       0.004673  0.087494      Dis   Dis\n",
       "M0909_1.02 M1009_1.02  0.315789       0.000000  0.085192      Dis   Dis\n",
       "           M1010_1.02  0.315789       0.000000  0.085192      Dis   Dis\n",
       "           M0966_1.02  0.280702       0.007576  0.063947      Dis   Dis\n",
       "M0933_1.02 M1049_1.02  0.400000       0.009346  0.054308      Dis   Dis\n",
       "M0931_1.02 M0933_1.02  0.290909       0.000000  0.054079      Dis   Dis\n",
       "M0933_1.02 M0941_1.02  0.280702       0.000000  0.049819      Dis   Dis\n",
       "M0909_1.02 M1007_1.02  0.263158       0.000000  0.049280      Dis   Dis\n",
       "           M1049_1.02  0.381818       0.000000  0.043178      Dis   Dis\n",
       "           M0941_1.02  0.280702       0.000000  0.042790      Dis   Dis\n",
       "M0933_1.02 M1073_1.02  0.192982       0.004673  0.042592      Dis   Dis\n",
       "           M1072_1.02  0.192982       0.004673  0.042592      Dis   Dis\n",
       "M0909_1.02 M1072_1.02  0.245614       0.000000  0.036407      Dis   Dis\n",
       "           M1073_1.02  0.245614       0.000000  0.036407      Dis   Dis\n",
       "           M0931_1.02  0.263158       0.009390  0.023920      Dis   Dis\n",
       "M0931_1.02 M1049_1.02  0.309091       0.000000 -0.027452      Dis   Dis\n",
       "M1009_1.02 M1049_1.02  0.272727       0.000000 -0.027919      Dis   Dis\n",
       "M1010_1.02 M1049_1.02  0.272727       0.000000 -0.027919      Dis   Dis\n",
       "M1007_1.02 M1049_1.02  0.218182       0.000000 -0.036610      Dis   Dis\n",
       "M1049_1.02 M1072_1.02  0.228070       0.000000 -0.039781      Dis   Dis\n",
       "           M1073_1.02  0.228070       0.000000 -0.039781      Dis   Dis\n",
       "M0966_1.02 M1049_1.02  0.290909       0.000000 -0.045417      Dis   Dis"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Read SR Model\n",
    "HD_SRModel = srpred.ReadSRModel('../SRModels/F223_1.97d.json')\n",
    "\n",
    "#Score Motif Alignments\n",
    "SR_Scores_i = []\n",
    "SR_Scores = []\n",
    "for pair, sr_alignment in MotifAlignments.items():\n",
    "    SR_Score, SR_Class = srpred.ScoreAlignmentResult(resultDict=sr_alignment, scoreDict=HD_SRModel)\n",
    "    SR_Scores_i.append(pair)\n",
    "    SR_Scores.append([sr_alignment['PctID_L'], sr_alignment['EScoreOverlap'], SR_Score, SR_Class])\n",
    "SR_Scores = pd.DataFrame(SR_Scores, columns = ['AA %ID','EScoreOverlap', 'SR_Score', 'SR_Class'])\n",
    "SR_Scores.index = pd.MultiIndex.from_tuples(SR_Scores_i)\n",
    "SR_Scores['Class'] = 'Amb'\n",
    "SR_Scores.loc[SR_Scores['EScoreOverlap'] >= Homeodomain_ReplicateThreshold, 'Class'] = 'HSim'\n",
    "SR_Scores.loc[SR_Scores['EScoreOverlap'] < 0.2, 'Class'] = 'Dis'\n",
    "SR_Scores.sort_values('SR_Score', ascending=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
